crates/holochain_types/src/validation_receipt.rs

```diff
@@ -0,0 +1,151 @@
+
+use crate::prelude::{Signature, Timestamp};
+use futures::{Stream, StreamExt, TryStreamExt};
+use holo_hash::{AgentPubKey, DhtOpHash};
+use holochain_keystore::{AgentPubKeyExt, MetaLairClient};
+use holochain_serialized_bytes::prelude::*;
+use holochain_zome_types::ValidationStatus;
+use std::vec::IntoIter;
+
+/// Validation receipt content - to be signed.
+#[derive(
+    Debug,
+    Clone,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Hash,
+    serde::Serialize,
+    serde::Deserialize,
+    SerializedBytes,
+)]
+pub struct ValidationReceipt {
+    /// the op this validation receipt is for.
+    pub dht_op_hash: DhtOpHash,
+
+    /// the result of this validation.
+    pub validation_status: ValidationStatus,
+
+    /// the remote validator which is signing this receipt.
+    pub validators: Vec<AgentPubKey>,
+
+    /// Time when the op was integrated
+    pub when_integrated: Timestamp,
+}
+
+impl ValidationReceipt {
+    /// Sign this validation receipt.
+    pub async fn sign(
+        self,
+        keystore: &MetaLairClient,
+    ) -> holochain_keystore::LairResult<Option<SignedValidationReceipt>> {
+        if self.validators.is_empty() {
+            return Ok(None);
+        }
+        let this = self.clone();
+        // Try to sign with all validators but silently fail on
+        // any that cannot sign.
+        // If all signatures fail then return an error.
+        let futures = self
+            .validators
+            .iter()
+            .map(|validator| {
+                let this = this.clone();
+                let validator = validator.clone();
+                let keystore = keystore.clone();
+                async move { validator.sign(&keystore, this).await }
+            })
+            .collect::<Vec<_>>();
+        let stream = futures::stream::iter(futures);
+        let signatures = try_stream_of_results(stream).await?;
+        if signatures.is_empty() {
+            unreachable!("Signatures cannot be empty because the validators vec is not empty");
+        }
+        Ok(Some(SignedValidationReceipt {
+            receipt: self,
+            validators_signatures: signatures,
+        }))
+    }
+}
+
+/// Try to collect a stream of futures that return results into a vec.
+async fn try_stream_of_results<T, U, E>(stream: U) -> Result<Vec<T>, E>
+where
+    U: Stream,
+    <U as Stream>::Item: futures::Future<Output = Result<T, E>>,
+{
+    stream.buffer_unordered(10).map(|r| r).try_collect().await
+}
+
+/// A full, signed validation receipt.
+#[derive(
+    Debug,
+    Clone,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Hash,
+    serde::Serialize,
+    serde::Deserialize,
+    SerializedBytes,
+)]
+pub struct SignedValidationReceipt {
+    /// the content of the validation receipt.
+    pub receipt: ValidationReceipt,
+
+    /// the signature of the remote validator.
+    pub validators_signatures: Vec<Signature>,
```

Comment:
Uhh... this isn't something you introduced, but we're not including the encoded bytes, so there could be problems validating the signatures, since messagepack isn't strictly deterministic.

Reply:
This `Signature` is a newtype around a byte array - `pub struct Signature(pub [u8; SIGNATURE_BYTES]);`

Is that not enough?

Reply:
I assume that is the 64 bytes of a detached signature that doesn't contain the original message. If it is a full signature containing the original message, then it would indeed be enough.

Reply:
Not asking you to do anything about this right now. I'm not sure we even validate validation receipts... we just count them. Maybe add a todo?

Reply:
Ahh I see, yes it is just 64 bytes. I'll add a TODO 👍🏻